From 03f5665e4a837b07b683b1a1217859daf3ec1b65 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:50 -0400
Subject: [fs] nfsd: Migrate the stateid reference into nfs4_find_stateid_by_type()

Message-id: <1410898849-22126-107-git-send-email-bfields@redhat.com>
Patchwork-id: 95821
O-Subject: [Patch RHEL7 106/165] nfsd: Migrate the stateid reference into nfs4_find_stateid_by_type()
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Allow nfs4_find_stateid_by_type to take the stateid reference, while
still holding the &cl->cl_lock. Necessary step toward client_mutex
removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 2d3f96689ffc757628c6d4038cacaaeb72a03345
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index a4a49a3..653de6b 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -1696,8 +1696,12 @@ find_stateid_by_type(struct nfs4_client *cl, stateid_t *t, char typemask)
 
  spin_lock(&cl->cl_lock);
  s = find_stateid_locked(cl, t);
- if (s != NULL && !(typemask & s->sc_type))
-  s = NULL;
+ if (s != NULL) {
+  if (typemask & s->sc_type)
+   atomic_inc(&s->sc_count);
+  else
+   s = NULL;
+ }
  spin_unlock(&cl->cl_lock);
  return s;
 }
@@ -3326,8 +3330,6 @@ static struct nfs4_delegation *find_deleg_stateid(struct nfs4_client *cl, statei
  ret = find_stateid_by_type(cl, s, NFS4_DELEG_STID);
  if (!ret)
   return NULL;
- /* FIXME: move into find_stateid_by_type */
- atomic_inc(&ret->sc_count);
  return delegstateid(ret);
 }
 
@@ -4170,8 +4172,6 @@ nfsd4_lookup_stateid(struct nfsd4_compound_state *cstate,
  *s = find_stateid_by_type(cstate->clp, stateid, typemask);
  if (!*s)
   return nfserr_bad_stateid;
- /* FIXME: move into find_stateid_by_type */
- atomic_inc(&(*s)->sc_count);
  return nfs_ok;
 }
 
-- 
1.7.1